### PR TITLE
Add an `attach` function to `ParamScheduler`.

### LINKED ISSUE #2291: Attach method for ParamScheduler
## 🚀 Feature
`StateParamScheduler`has introduced an `attach` method. Following this comment https://github.com/pytorch/ignite/pull/2090#discussion_r717412539 , maybe this could also be introduced in the optimizer's parameter scheduler (`ParamScheduler`) ? It will be necessary to keep BC.

### PR DESCRIPTION
**Pull request by atharva7-g**
Fixes #2291.

Description:

Added an attach function, have not edited save history within the attach function since it is lazily handled in `__call__`. I have also added a parametrized test to `tests/ignite/handlers/test_param_scheduler.py`. 

I have added tests for all subclasses, have modified tests with a `use_attach` parameter. Please let me know if the code, the tests and the docstring look OK. 

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/handlers/param_scheduler.py [hunk 1]
```diff
@@ -192,6 +192,20 @@ def __init__(
...
+        self,
+        engine: Engine,
+        event: str | Events | CallableEventWithFilter | EventsList = Events.ITERATION_STARTED,
+    ) -> None:
+        """Attach the handler to the engine.
+
+        Args:
+            engine: trainer to which the handler will be attached.
+            event: trigger event to update the param value.
+
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:205]: Let's add `.. versionadded:: 0.5.4` here
```suggestion
        .. versionadded:: 0.5.4
```

See this: https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md#writing-documentation

File: ignite/handlers/param_scheduler.py [hunk 2]
```diff
@@ -192,6 +192,20 @@ def __init__(
...
+        self,
+        engine: Engine,
+        event: str | Events | CallableEventWithFilter | EventsList = Events.ITERATION_STARTED,
+    ) -> None:
+        """Attach the handler to the engine.
+
+        Args:
+            engine: trainer to which the handler will be attached.
+            event: trigger event to update the param value.
+        .. versionadded:: 0.4.7
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py]:
```suggestion

        .. versionadded:: 0.5.4
```

File: tests/ignite/handlers/test_param_scheduler.py [hunk 3]
```diff
@@ -1451,3 +1451,116 @@ def get_cos_shed():
...
+    [
+        (LinearCyclicalScheduler, {"param_name": "lr", "start_value": 1.0, "end_value": 0.0, "cycle_size": 10}),
+        (CosineAnnealingScheduler, {"param_name": "lr", "start_value": 0.0, "end_value": 1.0, "cycle_size": 10}),
+        (
+            PiecewiseLinear,
+            {"param_name": "lr", "milestones_values": [(5, 0.5), (15, 1.0), (25, 0.0), (35, 1.0), (40, 0.5)]},
+        ),
+    ],
+)
+def test_param_scheduler_attach_equivalence(scheduler_cls, kwargs):
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_param_scheduler.py]: Concerning the tests, let's add `use_attach` arg to the existing tests and update the code instead of checking consistency as you do here.
Thanks !

Comment (atharva7-g) [reply tests/ignite/handlers/test_param_scheduler.py]: Okay, fair. My only concern was that rewriting existing tests might become a bit too complicated since I would have to make a sub function with a `use_attach` argument for every test. The tests I have written right now should be working fine as-it-is.

But I'll take a look.

Thanks!

File: tests/ignite/handlers/test_param_scheduler.py [hunk 4]
```diff
@@ -79,78 +79,88 @@ def test_linear_scheduler_asserts():
...
 def test_linear_scheduler():
-    tensor = torch.zeros([1], requires_grad=True)
-    optimizer = torch.optim.SGD([tensor], lr=0.0)
-
-    scheduler = LinearCyclicalScheduler(optimizer, "lr", 1, 0, 10)
-    state_dict = scheduler.state_dict()
-
-    def save_lr(engine):
-        lrs.append(optimizer.param_groups[0]["lr"])
+    def run(use_attach):
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_param_scheduler.py]: Let's use pytest parametrize instead of inner `run` function:
```python
@pytest.mark.parametrize("use_attach", [True, False])
def test_linear_scheduler(use_attach):
```

File: tests/ignite/handlers/test_param_scheduler.py [hunk 5]
```diff
@@ -282,51 +292,58 @@ def save_lr(engine):
 
 @pytest.mark.parametrize("cyclic_warmup", [False, True])
 def test_cosine_annealing_scheduler(cyclic_warmup):
-    tensor = torch.zeros([1], requires_grad=True)
-    optimizer = torch.optim.SGD([tensor], lr=0)
+    def run(use_attach):
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_param_scheduler.py]: Same here and everywhere you introduced `def run(use_attach)`

Comment (atharva7-g) [reply tests/ignite/handlers/test_param_scheduler.py]: Oh fair enough, actually. Please lmk if it looks OK now.
